server/src/main/java/org/elasticsearch/index/engine/SoftDeletesPolicy.java

```diff
@@ -114,7 +127,20 @@ synchronized long getMinRetainedSeqNo() {
      * Returns a soft-deletes retention query that will be used in {@link org.apache.lucene.index.SoftDeletesRetentionMergePolicy}
      * Documents including tombstones are soft-deleted and matched this query will be retained and won't cleaned up by merges.
      */
-    Query getRetentionQuery() {
-        return LongPoint.newRangeQuery(SeqNoFieldMapper.NAME, getMinRetainedSeqNo(), Long.MAX_VALUE);
+    synchronized Query getRetentionQuery() {
+        final BooleanQuery.Builder queryBuilder = new BooleanQuery.Builder();
+        queryBuilder.add(LongPoint.newRangeQuery(SeqNoFieldMapper.NAME, getMinRetainedSeqNo(), Long.MAX_VALUE), BooleanClause.Occur.SHOULD);
+        if (retentionAge.seconds() > 0) {
+            // we prefer using a docValues instead of an indexed field to avoid impact on append-only indices.
+            // however, it's a trade-off as we have to linearly scan to find the most recent documents.
+            final Query timestampQuery = NumericDocValuesField.newSlowRangeQuery(SeqNoFieldMapper.UPDATE_TIMESTAMP_IN_SECONDS_NAME,
```

Comment:
I think this the right trade off here.

Reply:
Can you explain that a bit more? If there are no soft deleted documents in a segment, the overhead of an indexed field is 0? also, the overhead is linear in the number of soft-deleted documents? The upside of an indexed field is that it's much faster to query to see if we need to merge a segment away.

Reply:
@bleskes I added a comment.

Reply:
Thanks @dnhatn. From you comment I read that the main point here is that we need to update existing docs with this value (with soft deletes) or store it there originally  (with an index field).

Reply:
@bleskes Good point. I'll add it.